src/bin/trig.rs

```diff
@@ -461,14 +537,159 @@ impl SynthLanguage for Math {
 
     fn validate(
         _synth: &mut Synthesizer<Self>,
-        _lhs: &Pattern<Self>,
-        _rhs: &Pattern<Self>,
+        lhs: &Pattern<Self>,
+        rhs: &Pattern<Self>,
     ) -> ValidationResult {
-        ValidationResult::Valid
+        let valid_pattern = |pat: &Pattern<Self>| {
+            pat.ast.as_ref().iter().all(|n| match n {
+                ENodeOrVar::ENode(Math::Div([_, j])) => match pat.ast.index(*j) {
+                    ENodeOrVar::Var(_) => true,
+                    ENodeOrVar::ENode(n) => !is_zero(n),
+                },
+                _ => true,
+            })
+        };
+
+        ValidationResult::from(valid_pattern(lhs) && valid_pattern(rhs))
+    }
+
+    fn is_allowed_rewrite(lhs: &Pattern<Self>, rhs: &Pattern<Self>) -> bool {
+        let contains_trig_node = |pat: &Pattern<Self>| {
+            pat.ast.as_ref().iter().any(|n| {
+                matches!(
+                    n,
+                    ENodeOrVar::ENode(Math::Sin(_))
+                        | ENodeOrVar::ENode(Math::Cos(_))
+                        | ENodeOrVar::ENode(Math::Tan(_))
+                        | ENodeOrVar::ENode(Math::Csc(_))
+                        | ENodeOrVar::ENode(Math::Sec(_))
+                        | ENodeOrVar::ENode(Math::Cot(_))
+                )
+            })
+        };
+
+        let pattern_is_extractable = |pat: &Pattern<Self>| {
+            pat.ast.as_ref().iter().all(|n| match n {
+                ENodeOrVar::Var(_) => true,
+                ENodeOrVar::ENode(n) => n.is_extractable(),
+            })
+        };
+
+        (contains_trig_node(lhs) || contains_trig_node(rhs))
+            && pattern_is_extractable(lhs)
+            && pattern_is_extractable(rhs)
+    }
+
+    fn is_valid_rewrite(
```

Comment:
what's this for?

Reply:
This is for filtering what rules we would like to see. The rule must contain extractable ops and there must be at least one trig op on the left-hand or right-hand side of the rule.